File: VB/Applications/AvaloniaLoggingVBNoDI/Extensions/LoggerExtension.vb
Imports LogViewerVB.Core
Imports Microsoft.Extensions.DependencyInjection
Imports Microsoft.Extensions.Logging
Imports Microsoft.Extensions.Logging.Configuration
Imports System.Runtime.CompilerServices
Imports AvaloniaLoggingVBNoDI.DataStores
Imports Microsoft.Extensions.DependencyInjection.Extensions
Imports MsLoggerVB.Core

Public Module LoggerExtension

    <Extension>
    Public Function AddDataStoreLogger(builder As ILoggingBuilder) As ILoggingBuilder

        builder.AddConfiguration()

        ' We need to use a shared instance of the DataStore to pass to the LogViewerControl
        builder.Services.AddSingleton(MainControlsDataStore.DataStore)
        builder.Services.TryAddEnumerable(ServiceDescriptor.Singleton(Of ILoggerProvider, DataStoreLoggerProvider))
        Return builder

    End Function

    <Extension>
    Public Function AddDataStoreLogger(builder As ILoggingBuilder, configure As Action(Of DataStoreLoggerConfiguration)) As ILoggingBuilder

        builder.AddDataStoreLogger()
        builder.Services.Configure(configure)
        Return builder

    End Function

End Module

File: CSharp/Applications/WpfLoggingNoDI/Extensions/LoggerExtension.cs
using System;
using LogViewer.Core;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.DependencyInjection.Extensions;
using Microsoft.Extensions.Logging;
using Microsoft.Extensions.Logging.Configuration;
using MsLogger.Core;
using WpfLoggingNoDI.DataStores;

namespace WpfLoggingNoDI.Extensions;

public static class LoggerExtension
{
    public static ILoggingBuilder AddDataStoreLogger(this ILoggingBuilder builder)
    {
        builder.AddConfiguration();

        // We need to use a shared instance of the DataStore to pass to the LogViewerControl
        builder.Services.AddSingleton(MainControlsDataStore.DataStore);
        builder.Services.TryAddEnumerable(ServiceDescriptor.Singleton<ILoggerProvider, DataStoreLoggerProvider>());
        return builder;
    }

    public static ILoggingBuilder AddDataStoreLogger(this ILoggingBuilder builder, Action<DataStoreLoggerConfiguration> configure)
    {
        builder.AddDataStoreLogger();
        builder.Services.Configure(configure);
        return builder;
    }
}